examples/nucleo_f429zi/usbfatfs/main.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2020, Niklas Hauser
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include <fatfs/ff.h>
+#include <tusb.h>
+#include <modm/board.hpp>
+#include <modm/processing.hpp>
+#include <modm/math/utils.hpp>
+
+/* You can test this by updating this firmware with itself:
+
+If you want to use the sanity check:
+
+	scons bin && \
+	cp ../../../build/nucleo_f429zi/usb_fatfs/scons-release/usbfatfs.bin . && \
+	echo -n "F429" >> usbfatfs.bin && \
+	crc32 usbfatfs.bin | xxd -r -p - >> usbfatfs.bin && \
+	mv usbfatfs.bin /Volumes/MODM_USB
+
+Otherwise just copy the file as is and `#define WITH_SANITY_CHECK 0`
+
+	scons bin && \
+	cp ../../../build/nucleo_f429zi/usb_fatfs/scons-release/usbfatfs.bin /Volumes/MODM_USB
+
+*/
+
+#define WITH_SANITY_CHECK 1
+
+// ----------------------------------------------------------------------------
+modm_ramcode
+void ram_apply(uint8_t pages, const uint8_t *image, uint32_t length)
+{
+	modm::atomic::Lock l;
+	// ONLY RAMCODE FROM HERE ON!!!
+	for (auto page{0u}; page <= pages; page++)
+		Flash::erase(page);
+	for (uintptr_t ptr{0}; ptr < length; ptr += sizeof(Flash::MaxWordType))
+		Flash::program(Flash::OriginAddr + ptr, *(Flash::MaxWordType*)(image + ptr));
```

Comment:
Deferencing `*(Flash::MaxWordType*)(image + ptr)` is UB unless `Flash::MaxWordType` is one of `char`, `unsigned char`, `signed char` or `std::byte`. The standard requires you to use `memcpy` here which would get optimized out. At least until we get `std::start_lifetime_as` from C++23. If it works we can leave it as is for now and fix it later.